### PR TITLE
[TRIVIAL IMPORTANT] first fixes for static __ctfe

### PR DESCRIPTION
**Pull request by UplinkCoder**
Hi this is a small fix deblocking  https://github.com/dlang/dmd/pull/6098

### CODE REVIEW

File: src/rt/util/hash.d [hunk 1]
```diff
@@ -33,10 +33,14 @@ size_t hashOf( const(void)[] buf, size_t seed )
         version( HasUnalignedOps )
         {
             if (!__ctfe)
+            {
                 return *cast(ushort*) x;
+            }
```
Comment (schveiguy) [inline src/rt/util/hash.d]: In the case of `version(HasUnalignedOps)`, and `__ctfe == true`, this function has no body.

**Comment by schveiguy** [comment]
I'm a bit puzzled. Why is this needed? Is it because of constant folding?

**Comment by UplinkCoder** [comment]
@schveiguy yes. 
Otherwise the code will be flagged as unreachable.

**Comment by schveiguy** [comment]
Hm... the non-DRY-ness of this looks awful. I suppose this is necessary to work with both compile time and runtime `__ctfe` definitions.

**Comment by UplinkCoder** [comment]
@schveiguy one the static __ctfe goes in.
It can be pretty!

**Comment by UplinkCoder** [comment]
@WalterBright or @andralex please merge.

**Comment by schveiguy** [comment]
I think this may be a big problem. All current code uses `if(__ctfe)`, and if this is not just a drop-in replacement, we are talking about breaking a lot of code for zero benefit. If this PR is any indication, the "fix" is not trivial.

Regardless of the obvious benefits, I think your dmd patch needs to continue to treat `__ctfe` as a magic runtime variable when used at runtime, and a compile-time constant when used at compile time.

**Comment by UplinkCoder** [comment]
The user would get a depreciation warning about unreachable code.
And that warning is totally legitimate.

**Comment by UplinkCoder** [comment]
The only reason this kludge is needed is because the d-auto-tester compiles with -w.
And the dmd pr is not yet in.
once the pr is in the fix it to change all if(__ctfe) to static if (__ctfe) done!

**Comment by schveiguy** [comment]
> The only reason this kludge is needed is because the d-auto-tester compiles with -w.

And so do many many projects. -w finds bugs.

> And that warning is totally legitimate.

That's like saying a function like this:

```D
char foo(bool flag)
{
    if(flag)
        return 'y';
    return 'n';
}
```

when called via `foo(true)` and -inline should flag a "legitimate" unreachable code error. It's useless noise.

**Comment by UplinkCoder** [comment]
Yes you are right. 
However that is a different problem. 
And one which has been brought up before.

**Comment by schveiguy** [comment]
It is not a different problem. The code is reachable, just not with that specific call. I happen to think actually that we are too strict on that requirement. For example, a template with a boolean should not trigger unreachable code -- sometimes the boolean is true, sometimes false. It shouldn't say the code is unreachable, just because it's unreachable in a certain instantiation.

But in this case, you are going to flag lots of existing code that works perfectly fine and is *understood* to be unreachable at compile time or at runtime.

I'm wondering if `__ctfeStatic` or something would be better to introduce, so as not to break existing code.

**Comment by andralex** [comment]
I agree this could be problematic. I routinely mark code that uses the redundant `else` in code reviews.

**Comment by UplinkCoder** [comment]
With dmd's current control-flow-analysis. The else is not redundant. 
I thought about `__ctfeStatic` and discarded it. 
We need to clean the unreachable situation up. 
Not introduce hacks.

**Comment by jmdavis** [comment]
> With dmd's current control-flow-analysis. The else is not redundant. 

As long as you're dealing with runtime checks, if the if returns, the else is redundant, and it's useless noise to have it. It's only with static if that it makes sense, because it actually affects which code is compiled in and not just what is run.  `__ctfe` is a weird beast, because it's affecting what gets run at compile time, but it's still affecting what gets run and not what gets compiled in.

> The only reason this kludge is needed is because the d-auto-tester compiles with -w.

How about making `__ctfe` work in both static if statements and if statements where it has the current behavior in if statements and the new behavior in static if statements and then we deprecate the use of `__ctfe` as a runtime variable. Then you don't run afoul of -w, since deprecation is separate from warnings. Anyone using `__ctfe` will be bugged to update their code, but their code won't break. And once the deprecation is done with, you'll just have `__ctfe`, so it will be clean like you're looking for, but it won't break code in the interim.

**Comment by UplinkCoder** [comment]
@klickverbot making __ctfe a CT-Entity, is fairly useful.
please elaborate on why it is a mistake,

**Comment by andralex** [comment]
@UplinkCoder to make sure I understand, consider a library with the code:

```D
string fun(int x)
{
    if (x < 0) return "negative";
    ... a bunch of code ...
    return "positive";
}
```

Now, say in a particular program `fun` is evaluated during compilation and only with negative values. Does your code introduce a warning about unreachable code?

**Comment by UplinkCoder** [comment]
No not at all.
All my code does is to change `__ctfe` from a variable into a literal. 
However that does not work with the way function are currently handled.

**Pull request closed without merging**